### ISSUE TITLE
Consider adding sort key in RewriteManifests

### ISSUE THREAD

**Issue by aokolnychyi** [apache/incubator-iceberg#481]
In one of our use cases, we write to many partitions within every batch. As a result, each manifest covers a lot of partitions, which slows down the job planning.

We decided to use `RewriteManifests` to optimize the layout of the metadata. Our idea was to group files based on their partition values. However, there is a problem. Let's say we have a table with 3 partition columns. If I cluster metadata by all of them, this will produce small manifest files (especially for partitions that don't have all its data yet). I don't want to have too granular manifests but I want all entries for my leaf partitions to be next to each other. If we cluster metadata let's say by the first partition value, then entries for leaf partitions are not next to each other, which slows down the job planning also.

The example above made me think about introducing a sort key in addition to the cluster key so that entries are grouped by that sort key within clusters. The API can look like:

```
public interface RewriteManifests<C, S>  extends SnapshotUpdate<RewriteManifests<C, S>> {
  RewriteManifests clusterBy(BiFunction<PartitionSpec, DataFile, C> func);
  RewriteManifests sortBy(BiFunction<PartitionSpec, DataFile, S> func, Comparator<S> cmp);
  RewriteManifests rewriteIf(Predicate<ManifestFile> predicate);
}
```

The problem with the above suggestion is that we will need to keep all the files in memory before writing them to manifests. Metadata rewrites can be expensive and we need a distributed version of it but I am wondering about the overall idea for now. It is clear that such an approach can work only if we compact only a part of the metadata at once.

**Comment by aokolnychyi** [comment]
@bryanck @rdblue what are your thoughts?

**Comment by bryanck** [comment]
Would clustering by a range work instead of sorting? E.g. cluster by the first partition column and ranges of the 2nd? You could make one pass through the files to determine the ranges if they aren't evenly distributed.

**Comment by rdblue** [comment]
How does sorting help? I don't think there is anything that understands that a manifest may be sorted, so the cost of planning would be the same for sorted and unsorted.

I agree with Bryan that it probably makes the most sense to come up with partition ranges that give you the right size of manifest files for the table.

**Comment by aokolnychyi** [comment]
Let’s consider a specific example: we have a table partitioned by col1 and further bucketed into 20 buckets by col2. On one hand, clustering by both columns is too granular and yields small manifests. Alternatively, clustering only by col1 would mean producing let’s say 5 manifests of the needed size for each top level partition. The problem is that each of those 5 manifests can contain files for the same leaf partition. As a result, we will be scanning all the metadata for the top level partition even though we query for a specific leaf partition. 

To address those issues, I had multiple approaches in mind. 

Option 1

- Estimate the size of one manifest entry by iterating through the metadata for manifests we are rewriting.
- Read manifests and cluster the files.
- Compute the number of files per group and estimate the size of metadata. 
- Use bin packing on the info from the step above assuming the clustering key is sortable. That way, we will group metadata for leaf partitions together. 
- Finally, group files based on bins. 

Option 2

- Introduce a sort key in addition to the clustering key. In the example above, we can say cluster data by col1 and sort by col2.
- Read manifests and cluster files. 
- Sort files within groups. 
- Write manifests respecting the target manifest size. In our case, we will produce 5 manifests but only one of them will contain data for the same leaf partition (unless we were unlucky and we closed the manifest in the middle, which we can prevent). 

To sum up, sort key would only help if we also respect the target manifest size.

I considered these approaches from a distributed job perspective and it seemed we would need to scan the metadata multiple times using the first option. We can cache, of course.

@bryanck @rdblue is my first option similar to what you suggest?

**Comment by rdblue** [comment]
Okay, I understand the motivation for sorting now. I think sorting is still a difficult way to solve this problem because you'd need to keep so many files in memory. Tables that need metadata rewrites usually have lots of data files.

The approach we took is to look at the metadata tables to get the number of entries and the current size of all manifests, and then use that to estimate how many buckets should be grouped together. Then the grouping function returns a key with `bucket_num % group_size`.

That's a much simpler approach than sorting. The drawback is that it relies on the number of files per bucket staying fairly stable, but I think that's reasonable in most cases.

**Comment by rdblue** [comment]
@aokolnychyi, I don't think sorting is the right solution. Would you mind if we closed this issue?

**Comment by aokolnychyi** [comment]
Yeah, I went for option 1 as described in my previous reply. Let's close this one and I will create a few others based on what we discussed.
